From 9ed0c5a7308bdba41d27f5726f8516aa3fb97dbc Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Mon, 4 Sep 2017 12:03:42 +0800
Subject: [PATCH 2496/5242] MLK-16353-3 video: fbdev: dcss: add alignment
 check in check_var()

commit  065b9941a808c10fa23ca9ac8254fea4f949577d from
https://source.codeaurora.org/external/imx/linux-imx.git

Add alignment check for SCALER which requires both
width and height to be divisable by 4. Put the check
in fb_check_var hook.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/imx_dcss.c |   21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/video/fbdev/mxc/imx_dcss.c b/drivers/video/fbdev/mxc/imx_dcss.c
index d6b21ec..f60474a 100644
--- a/drivers/video/fbdev/mxc/imx_dcss.c
+++ b/drivers/video/fbdev/mxc/imx_dcss.c
@@ -2429,6 +2429,27 @@ static int dcss_check_var(struct fb_var_screeninfo *var,
 		}
 		var->bits_per_pixel = format->bpp;
 	}
+
+	/* Add alignment check for scaler */
+	switch (fmt_is_yuv(var->grayscale)) {
+	case 0:         /* ARGB8888 */
+		if (ALIGN(var->xres, 4) != var->xres ||
+		    ALIGN(var->yres, 4) != var->yres) {
+			dev_err(&pdev->dev, "width or height is not aligned\n");
+			return -EINVAL;
+		}
+		break;
+	case 2:         /* YUV420 */
+		if (ALIGN(var->xres, 16) != var->xres ||
+		    ALIGN(var->yres, 8) != var->yres) {
+			dev_err(&pdev->dev, "width or height is not aligned\n");
+			return -EINVAL;
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+
 	fix->line_length = var->xres * (var->bits_per_pixel >> 3);
 	fb_size = var->yres_virtual * fix->line_length;
 
-- 
1.7.9.5

